### DISCUSSION TITLE
Can't get referrer in NextJS Middleware next 12.3.4

### DISCUSSION THREAD

**Discussion by dubbajubba** [vercel/next.js#43352] [Help] [answered]
### Describe the Bug
I'm trying to access the referrer in my NextJS middleware and it just returns `undefined`. My site is deployed on Vercel and I'm running `next v12.3.4`.

Here is my test middleware code:
```typescript
export default async function slugMiddleware( req: NextRequest ) {  
  console.log(req.referrer);
  return NextResponse.next();
}
```

Here is a screenshot from my Vercel functions log.
<img width="1277" alt="image" src="https://user-images.githubusercontent.com/33437206/203868791-c46a954e-ba5e-44d6-9523-176a0a9d54f5.png">

### Expected Behavior
I have shared a link on my Wordpress blog and followed the link there. I also messaged myself in Facebook and followed the link from there. I would expect the output of `req.referrer` to be my blog or Facebook, respectively.

Any help or suggestions would be appreciated. Thanks!

**Comment by icyJoseph** [comment, marked as answer]
Hi,

Do you observe the same via `req.headers['referer']`, or perhaps it is `req.headers.get('referer')`?


## Edit

I had typed referrer (correctly), but the correct header name is mispelled, so it should be `referer` ~  ¯\\_(ツ)_/¯

> Note: The original header name [Referer](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referer) is a misspelling of the word "referrer". The Referrer-Policy header does not share this misspelling. 

Try with `req.headers.get('referer')`

**Reply by dubbajubba** [reply]
Thanks @icyJoseph and @awareness481, `req.headers.get('referer')` does the job 👍 Do you know why the `req.referrer` property doesn't work, when `req.headers.get('referer')` does?

I have another question that I hope you can help with.

When looking at my Vercel function logs, I notice that the console is showing two logs from my GET request, one has the external referring site, and the other is the page itself. Is this behavior expected? I'm trying to understand why there is a second log with the destination page's URL.
![image](https://user-images.githubusercontent.com/33437206/204032367-9002fdd3-b23b-4189-9052-9e24f2e40aac.png)

**Reply by dubbajubba** [reply]
@icyJoseph, I know you're super busy helping answer everyone's questions in the discussion board, but I was wondering if you could shed some light on the question I have above.

I am trying to understand how the middleware is executed and why I am getting the referrer output to the log twice. This is actually causing me an issue because I try recording some analytics that include the referrer, but the referrer that is being recorded is always the page itself.

If this is expected behavior, I'd love to learn more, but I am having trouble finding documentation that dive's deep into how middleware handles a request. If this isn't expected behavior, I can review my code more to figure out what I may be doing wrong, but I feel like there isn't anything wrong with my code.

P.S. thanks so much for answering all my questions and the community's questions! Healthy community engagement is such an important part of a good product and you rock!

**Reply by icyJoseph** [reply]
It might be a JSON data request hitting the server, or a prefetching of data attempt through a next/link. You should be able to see this locally when building the app, if you deactivate data for this page I think it won't hit the middleware more than once.


This part with the JSON data requests, is a bit foggy, I think they call it pre-flight connection or something like that.

**Reply by dubbajubba** [reply]
Ok, gotcha, that makes sense! Thanks again for the help!! 🙏

**Comment by awareness481** [comment]
Should be `req.headers['referer']` AFAIK

**Reply by icyJoseph** [reply]
> Note: The original header name [Referer](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referer) is a misspelling of the word "referrer". The Referrer-Policy header does not share this misspelling.

Spot on... but 

![Screenshot 2022-11-25 at 09 11 32](https://user-images.githubusercontent.com/21013447/203932516-24b60bb3-e6bb-4bcc-9bab-d868becd2f48.png)

And I do think headers in middleware are an instance of Headers... so it is a Map under the hood with a `get` function.

**Reply by awareness481** [reply]
Oh yeah, `req.headers.get("referer")` seems to work ^^
